File: examples/ccxt.pro/js/build-ohlcv-many-symbols.js
'use strict';

const ccxt = require ('ccxt');
const asTable  = require ('as-table').configure ({ delimiter: ' | ' })

console.log ('CCXT Version:', ccxt.version)


async function loop (exchange, symbol, timeframe, completeCandlesOnly = false) {
    const durationInSeconds = exchange.parseTimeframe (timeframe)
    const durationInMs = durationInSeconds * 1000
    while (true) {
        try {
            const trades = await exchange.watchTrades (symbol)
            if (trades.length > 0) {
                const currentMinute = parseInt (exchange.milliseconds () / durationInMs)
                let ohlcvc = exchange.buildOHLCVC (trades, timeframe)
                if (completeCandlesOnly) {
                    ohlcvc = ohlcvc.filter (candle => parseInt (candle[0] / durationInMs) < currentMinute)
                }
                if (ohlcvc.length > 0) {
                    console.log("Symbol:", symbol, "timeframe:", timeframe);
                    console.log ('-----------------------------------------------------------')
                    console.log (asTable (ohlcvc))
                    console.log ('-----------------------------------------------------------')
                }
            }
        } catch (e) {
            console.log (symbol, e)
            // do nothing and retry on next loop iteration
            // throw e // uncomment to break all loops in case of an error in any one of them
            // break // you can also break just this one loop if it fails
        }
    }
}

async function main () {
    // select the exchange
     const exchange = new ccxt.pro.ftx ()

    if (exchange.has['watchTrades']) {
        await exchange.loadMarkets ()
        // Change this value accordingly
        const timeframe = '1m'

        const allSymbols = exchange.symbols;

        // arbitrary n symbols
        const limit = 5;
        // const selectedSymbols = allSymbols.slice(0, limit);
        // you can also specify the symbols manually
        // example:
        // const selectedSymbols = ['BTC/USDT', 'LTC/USDT']

        console.log(selectedSymbols);

        // Use this variable to choose if only complete candles 
        // should be considered
        const completeCandlesOnly = true

        await Promise.all (selectedSymbols.map (symbol => loop (exchange, symbol, timeframe, completeCandlesOnly)))

    } else {
        console.log (exchange.id, 'does not support watchTrades yet')
    }
}

main ()

File: examples/ccxt.pro/py/build-ohlcv-many-symbols.py
# -*- coding: utf-8 -*-

import asyncio
import ccxt.pro


print('CCXT Version:', ccxt.__version__)


async def loop(exchange, symbol, timeframe, complete_candles_only = False):
    duration_in_seconds = exchange.parse_timeframe(timeframe)
    duration_in_ms = duration_in_seconds * 1000
    while True:
        try:
            trades = await exchange.watch_trades(symbol)
            if len(trades) > 0:
                current_minute = int(exchange.milliseconds() / duration_in_ms)
                ohlcvc = exchange.build_ohlcvc(trades, timeframe)
                if complete_candles_only:
                    ohlcvc = [candle for candle in ohlcvc if int( candle[0] / duration_in_ms ) < current_minute]
                if (len(ohlcvc) > 0):
                    print ('-----------------------------------------------------------')
                    print("Symbol:", symbol, "timeframe:", timeframe)
                    print (ohlcvc)

        except Exception as e:
            print(str(e))
            # raise e  # uncomment to break all loops in case of an error in any one of them
            # break  # you can also break just this one loop if it fails


async def main():
    # select the exchange
    exchange = ccxt.pro.ftx()
    if exchange.has['watchTrades']:
        markets = await exchange.load_markets()
        # Change this value accordingly
        timeframe = '1m'
        limit = 5
        selected_symbols = list(markets.values())[:limit]
        # you can also specify the symbols manually
        # selected_symbols = ['BTC/USDT', 'ETH/USDT']

        # Use this variable to choose if only complete candles
        # should be considered
        complete_candles_only = True
        await asyncio.gather(*[loop(exchange, symbol, timeframe, complete_candles_only) for symbol in selected_symbols])
        await exchange.close()
    else:
        print(exchange.id, 'does not support watchTrades yet')


asyncio.run(main())
